### ISSUE TITLE
feature request: Dark mode

### ISSUE THREAD

**Issue by Altreus** [metacpan/metacpan-web#2182]
Many developers have their environments set to a dark background. Having metacpan be the only brilliant white window in one's entire desktop environment is ouchy.

An alternative CSS using a darker colour scheme (and the option to save the preference) would be supreme.

**Comment by clipland** [comment]
In case this materializes, please implement theme-ability in a way that doesn't require one to be logged in.

Also a "classic" theme would be nice (shameless plug: [just released add-on](/clipland/addon-cpan-theme-for-metacpan)).
Related issues: #1432,  #278, #1360

**Comment by oalders** [comment]
I think this completely acceptable, but I don't see any of the core devs working on this since it's pretty low on our list of stuff that seems broken.  I mean, I can't read their minds, but that's kind of my impression.

So, if someone else wants to do the heavy lifting, please do.  Maybe hit us up with a plan before you sink a lot of time into it. https://github.com/metacpan/metacpan-docker

**Comment by clipland** [comment]
@Altreus If you're really impatient, you may use the same approach our yesterday-released Firefox Add-On uses, client-side overrides.

Just go [there](/clipland/addon-cpan-theme-for-metacpan), fork and clone on github or just clone locally, and replace/copy into the core .js file what I've quickly hacked up _for you_ during break:

```// Override CSS rules with colors borrowed from Blender
var styles = "body, .autocomplete-suggestion { background-color: #727272 !important; color: #0a0a0a !important; }";
styles    += "a, a:hover, a:visited { color: #fa4; }";
styles    += ".navbar-default .navbar-nav > li > a { color: #da2037; }";
styles    += "li.active a { background-color:#727272 !important; border-top-color: #fa4 !important; box-shadow: inset -1px 0px 0 #000, inset 1px 0px 0 #000 !important; }";
styles    += "input { background: #b0b0b0 !important; background-image: linear-gradient(#999, #b0b0b0) !important; border-color: #393939 !important; color: #020202 !important; } ";
styles    += ".btn { background-image: linear-gradient(#a6a6a6, #8a8a8a) !important; border: 1px solid #393939; border-radius: 4px; color: #020202; text-shadow: none; }";
styles    += ".alert { background-color: #a6a6a6 !important; border-color: #b0b0b0 !important;  }";
styles    += ".navbar { background-color: #393939 !important; color: #fff !important; border-color: #000 !important; border-top-color: #000 !important; }";
styles    += ".content { border-left-color: #000 !important; }";
styles    += ".description { color: #161616 !important; }";
styles    += "pre { background-color: #b0b0b0 !important; }";

var styleSheet = document.createElement("style");
styleSheet.type = "text/css";
styleSheet.innerText = styles;
document.head.appendChild(styleSheet);
```

Then load it as temporary add-on or generate a self-signed .xpi and you'll have this here.
Color-scheme is based on fabulous [Blender](https://www.blender.org/).
![screenshot](https://user-images.githubusercontent.com/3533661/59864304-cca86700-9386-11e9-8d10-09551ccd446e.png)
If you're not using FF, but another browser (oh no, ...), it should be easily adaptable.

These overrides are admittedly hackish/heavy-handed and you might see flashes of the old white layout.
The overall approach to override the original stylesheets isn't so smart in general! Yet still, a client-side setting via a browser-extension/web-storage/localstorage might be ~~a good~~ the best way to implement a theme-switcher for MetaCPAN without requiring a login/cookies and being light on resources.

**Comment by CaseOf** [comment]
@clipland Do you think it can be set with a simple css theme for Stylus?

**Comment by clipland** [comment]
_Stylus_, the FF Add-On?
Well, I guess that's essentially the same. ;) Both solutions inject override CSS rules.

**Comment by ThisUsedToBeAnEmail** [comment]
volunteering and I've made a start over the past couple of days. 

<img width="917" alt="Screenshot 2020-10-01 at 09 11 44" src="https://user-images.githubusercontent.com/16151791/94784761-8a156d00-03c6-11eb-871d-fe44bc19b226.png">

<img width="1331" alt="Screenshot 2020-10-01 at 09 15 13" src="https://user-images.githubusercontent.com/16151791/94784849-a87b6880-03c6-11eb-9feb-28882aef8a5d.png">

**Comment by clipland** [comment]
Congrats! Nice color scheme. Even nicer: it seems configurable. Am I right in assuming (from first screencap) you came up with some sort of client-side config, relying on web-storage/localstorage/(browser-extension) - not requiring a login? Would be great.

**Comment by ThisUsedToBeAnEmail** [comment]
Yes, the plan is to make everything configurable, I also want to add functionality to change font size and family.

Currently I am using local storage to cache the account profile so that I don't need to make an API request per page refresh.

As I mentioned I am using an api and storing this against the author document. The reason I'm doing this is because I would like persistency between my devices, however I'll welcome more opinions on this.

**Comment by haarg** [comment]
I really don't think we should have this be configurable on a per color basis.  If someone wants to tweak individual colors, they can already do that with user stylesheets.

Having a dark mode seems reasonable.

**Comment by ThisUsedToBeAnEmail** [comment]
I'm not finished going through every page yet but currently you can configure the following colours: 

<img width="677" alt="Screenshot 2020-10-01 at 13 48 29" src="https://user-images.githubusercontent.com/16151791/94812039-15562900-03ee-11eb-978c-38651a9b0539.png">

I'm open to removing this and only having the dark mode switch, I would like some more opinions though.

Here are some more screenshots: 

<img width="1333" alt="Screenshot 2020-10-01 at 13 50 00" src="https://user-images.githubusercontent.com/16151791/94812322-7bdb4700-03ee-11eb-9b66-5ca9906c0cde.png">

![Uploading Screenshot 2020-10-01 at 13.51.36.png…]()

![Uploading Screenshot 2020-10-01 at 13.54.08.png…]()

**Comment by oalders** [comment]
That looks really good!

**Comment by konstantinos-giannakakis** [comment]
Hello! Any progress with this feature? Still could be pretty useful.

**Comment by oalders** [comment]
Thanks for the nudge. I've asked internally about what we need to do to get this merged.

**Comment by ThisUsedToBeAnEmail** [comment]
If you want something you can use now I wrote an extension:

https://chrome.google.com/webstore/detail/metacpantheme/dmdmhdilgmmdnklgdnokgfcfnbegfkil?hl=en
https://addons.mozilla.org/en-US/firefox/addon/meta-cpan-theme/

**Comment by clipland** [comment]
Note to anyone following/reading this: _please_ make it standards conformant and not reliant on logins/auth of some sort.

- Work with alternative style sheets. See this [Moz Dev post](https://developer.mozilla.org/en-US/docs/Web/CSS/Alternative_style_sheets) for details. Such alternate styles are then available in FF i.e. via "View > Page Style". Also, these style sheets can be triggered/selected by JS etc.

- Take a look into the [CSS prefers-color-scheme @media feature](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme). It allows a user to set a preference in browser so appropriate CSS styles get activated accordingly. To test this in FF, i.e., see [this post](https://developer.mozilla.org/en-US/docs/Tools/Page_Inspector/How_to/Examine_and_edit_CSS#view_media_rules_for_prefers-color-scheme).

**Comment by Altreus** [comment]
+1 for the prefers-color-scheme feature. I think browsers can read the system preferences on OSes that support it, and feed that through, so if you've got your OS set to dark mode the browser will ask the site to prefer dark mode.

**Comment by akarelas** [comment]
One of the few big sites that still don't support dark mode.

**Comment by oalders** [comment]
I would love to see dark mode as well, but someone needs to step up and volunteer the time.

**Comment by akarelas** [comment]
Even a simple "follow system's color mode" would be great - can be done with plain CSS

**Comment by oalders** [comment]
If there's a PR, we are happy to review it.

**Comment by akarelas** [comment]
I really think this should be set as first-priority. It's one of those things that separates neanderthal websites from active ones.

**Comment by oalders** [comment]
It looks like we should hold off on this until we upgrade from Bootstrap 3.
